### PR TITLE
Frontpage Extension

### LINKED ISSUE #48: Improve Extrinsics component to support multiple params
Currently the UI only supports a single input argument even though the runtime modules functions may expect a variable amount of arguments

### LINKED ISSUE #91: Frontend-template extension
This ticket is to keep track of a list of tasks on front-end template extension so that devhub tutorial/doc can refer to using this template instead of polkadot-js app when we need front-end interaction. 

This is because we have less control on polkadot-js app which may break integrating with Substrate node-template once in a while. We also gain better version control on which node- & front-end template work together.

### Extrinsics, after picking on a pallet and extrinsic function call:
- [x] detect # of params and each param type
- [x] allow submitting options (something in signedExtra)
- [x] allow option of sudo
- [x] allow option of submitting with signed or unsigned

### Chain State
- [x] allow query of pallet constants also
- [x] after choosing pallet storage, detect # of input options needed and determine each param type. Sometimes this could be 0. Now are have hard-coded to send one option out.

### LINKED ISSUE #92: Explore/Implement a module to submit Substrate RPC Call

### PR DESCRIPTION
**Pull request by jimmychu0807**
close #48
close #91 
close #92

This PR mainly rewrites `Extrinsics.js` (and rename to `Interactor.js`) and expands `TxButton.js` so that it is more versatile in handling 1) query, 2) extrinsic submission (with unsigned tx, signed tx, and sudo tx), 3) RPC, and 4) constants.

RPC query currently **ONLY** support functions that do not require parameters, as there is no meta data provided from the underlying API.

### CODE REVIEW

**Comment by jimmychu0807** [comment]
This is a quick demo of the frontend now:
https://www.awesomescreenshot.com/video/129116?key=ebb96ffd285c7e92014e4dd48092b66e

**Comment by jimmychu0807** [comment]
@JoshOrndorff @danforbes @riusricardo 

There are some edges I have not yet fully polished.
I think the video already show it doesn't unsubscribe for previous events in the pallet interactor. I will try to solve it in next PR. But this one seems stable enough to be merged into `master`.

Also feel free to file a ticket if you encounter any problem/bug when using this latest pallet interactor module.

Let me know what you think.

**Comment by jimmychu0807** [comment]
@riusricardo I rewrite your PR, so that now clicking on the user icon on the left of the account dropdown will copy the selected acct address.

File: src/AccountSelector.js [hunk 1]
```diff
@@ -66,13 +67,15 @@ function Main (props) {
...
-            color={accountSelected ? 'green' : 'red'}
-          />
+            : null }
+          <CopyToClipboard text={accountSelected}>
+            <Icon
+              name='users'
+              size='large'
+              circular
+              color={accountSelected ? 'green' : 'red'}
+            />
```
Comment (riusricardo) [inline src/AccountSelector.js:77]:
```suggestion
            <Button
              basic
              circular
              size='large'
              icon='users'
              color={accountSelected ? 'green' : 'red'}
            >
            </Button>
```

Comment (riusricardo) [reply src/AccountSelector.js:77]: Looks better if it is a button because it changes the cursor and becomes clickable :)

**Comment by danforbes** [comment]
This looks _awesome_ @jimmychu0807! One request: can we make it easier to provide `call` parameters to extrinsics? For instance, right now it appears that if I want to use the Sudo module to make a `sudo` call, I have to provide the hash of the function I want to call. Can we either change this so that it's like Apps UI where someone can select a function and provide the parameters _or_ provide a separate component that allows people to calculate the hash for a call? The former may be "nicer" but the latter may actually be better in this context :thinking:

**Comment by JoshOrndorff** [comment]
> RPC query currently ONLY support functions that do not require parameters, as there is no meta data provided from the underlying API.

True that there is no metadata for RPCs. But you can decorate the API with RPC information when creating the api instance. Here is an example https://github.com/substrate-developer-hub/recipes/blob/master/nodes/rpc-node/js/index.js#L9-L42

**Comment by danforbes** [comment]
@JoshOrndorff - good point, I used a bad example because it's easy to work around that particular case with the provided Sudo button. I still think we want it to be easier to provide `call` parameters for things like scheduled dispatches and multi-sig operations.

**Comment by danforbes** [comment]
@JoshOrndorff - looking at your other points though, I don't think that needs to be a blocker for this PR :+1:

**Comment by JoshOrndorff** [comment]
@danforbes I see what you mean now. It's when you explicitly select the "sudo" pallet.

I agree with you it would be good to support `Call` parameters. I also agree with you that it shouldn't block this PR. This is strictly better than what we have on `master` now.

So some follow-up issues:
* allow decorating RPC information in the ui config
* support `Call` paramaters
* support `Vec` parameters

**Comment by danforbes** [comment]
:100: Not only is it strictly better, it's more than sufficient to support our current tutorial use cases right now, which we decided was the goal we were after :rocket:

**Comment by jimmychu0807** [comment]
@riusricardo  yup, button looks better

@JoshOrndorff  @danforbes 
Thanks for the review and comments. So 3 followup items, there are also some issues about unsubscription of transaction events

A few questions:

1. > allow decorating RPC information in the ui config

    So this is done by the developer, and not some kind of auto-detection, right? I can look into this and track it in a separated ticket.

2. > support `Call` paramaters

    @danforbes I don't quite get this. What is it means to support `Call` parameter?

3. > support Vec parameters

    A quick way to do it is to accept a comma-separated string, and then break it when parsing it. This also remind me that currently the fn-t doesn't support `Option<>`. It requires every parameter to be filled.

**Comment by JoshOrndorff** [comment]
> So this is done by the developer, and not some kind of auto-detection, right?

Exactly. Just like the types.

**Comment by danforbes** [comment]
@jimmychu0807 - check out how Polkadot JS Apps UI handles `Call` type parameters to extrinsics such as `Sudo.sudo`; rather than just providing an input field to enter the encoded extrinsic call, it provides the user with dropdown fields to select the extrinsic they would like to call and nice fields for entering the arguments.

image

**Comment by jimmychu0807** [comment]
@danforbes  But this can just be done directly by specifying the pallet / callable and click the "sudo" button. `sudo()` function basically is just a wrapper on other pallet / callable & params, and is achieved with choosing the sudo action.

**Comment by danforbes** [comment]
True, but their are other extrinsics people may way to use that require `call` type parameters, like `multisig` or `schedule`.

**Pull request merged**